### PR TITLE
Fix compaction ut datarace

### LINKED ISSUE #15241: [Bug]: DataNode unittest datarace
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: latest
- Deployment mode(standalone or cluster):all
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

refer to this pr , https://github.com/milvus-io/milvus/pull/15237.
Its unittest failed.

[2022/01/17 04:01:32.329 +00:00] [INFO] [compaction_executor.go:80] ["start to execute compaction"] [planID=1]
[2022/01/17 04:01:32.329 +00:00] [WARN] [compaction_executor.go:97] ["compaction executor stop task"] [planID=1]
==================
WARNING: DATA RACE
Write at 0x00c00037e818 by goroutine 42:
  internal/race.Write()
      /usr/local/go/src/internal/race/race.go:41 +0x125
  sync.(*WaitGroup).Wait()
      /usr/local/go/src/sync/waitgroup.go:128 +0x126
  github.com/milvus-io/milvus/internal/datanode.(*mockCompactor).stop()
      /go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor_test.go:157 +0x8a
  github.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).stopTask()
      /go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:98 +0x204
  github.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).stopExecutingtaskByVChannelName.func1()
      /go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:112 +0x1b7
  sync.(*Map).Range()
      /usr/local/go/src/sync/map.go:345 +0x16b
  github.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).stopExecutingtaskByVChannelName()
      /go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:110 +0xe4
  github.com/milvus-io/milvus/internal/datanode.TestCompactionExecutor.func6()
      /go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor_test.go:107 +0x464
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1127 +0x202

Previous read at 0x00c00037e818 by goroutine 79:
  internal/race.Read()
      /usr/local/go/src/internal/race/race.go:37 +0x206
  sync.(*WaitGroup).Add()
      /usr/local/go/src/sync/waitgroup.go:71 +0x219
  github.com/milvus-io/milvus/internal/datanode.(*mockCompactor).compact()
      /go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor_test.go:138 +0x55
  github.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).executeTask()
      /go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:82 +0x2b2

Goroutine 42 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1178 +0x796
  github.com/milvus-io/milvus/internal/datanode.TestCompactionExecutor()
      /go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor_test.go:88 +0x148
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1127 +0x202

Goroutine 79 (running) created at:
  github.com/milvus-io/milvus/internal/datanode.(*compactionExecutor).start()
      /go/src/github.com/milvus-io/milvus/internal/datanode/compaction_executor.go:68 +0xc4
==================
[2022/01/17 04:01:32.329 +00:00] [WARN] [compaction_executor.go:84] ["compaction task failed"] [planID=1] [error="context canceled"]
[2022/01/17 04:01:32.330 +00:00] [INFO] [compaction_executor.go:91] ["end to execute compaction"] [planID=1]
[2022/01/17 04:01:32.330 +00:00] [WARN] [compaction_executor.go:114] [mock]

### Expected Behavior

_No response_

### Steps To Reproduce

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by XuanYang-cn**
Fixes: #15241

Signed-off-by: yangxuan <xuan.yang@zilliz.com>

/kind bug
/assign @congqixia @czs007

### CODE REVIEW

File: internal/datanode/compactor.go [hunk 1]
```diff
@@ -265,8 +278,11 @@ func (t *compactionTask) merge(mergeItr iterator, delta map[UniqueID]Timestamp,
 }
 
 func (t *compactionTask) compact() error {
-	t.wg.Add(1)
-	defer t.wg.Done()
+	if ok := funcutil.CheckCtxValid(t.ctx); !ok {
+		log.Error("compact wrong, task context done or timeout")
+		return errIllegalCompactionPlan
```
Comment (congqixia) [inline internal/datanode/compactor.go:283]: maybe should be the newly added `errContext`

File: internal/datanode/compaction_executor.go [hunk 2]
```diff
@@ -87,13 +103,12 @@ func (c *compactionExecutor) executeTask(task compactor) {
...
 }
 
 func (c *compactionExecutor) stopTask(planID UniqueID) {
-	task, loaded := c.executing.LoadAndDelete(planID)
-	if loaded {
+	task, ok := c.executing.Load(planID)
+	if ok {
 		log.Warn("compaction executor stop task", zap.Int64("planID", planID))
 		task.(compactor).stop()
 	}
```
Comment (congqixia) [inline internal/datanode/compaction_executor.go:114]: If `stop` invoked multiple times simultaneously , `waitgroup` might enter state of negative counter panic.

File: internal/datanode/compactor.go [hunk 3]
```diff
@@ -41,11 +43,14 @@ var (
...
 	errTransferType            = errors.New("transfer intferface to type wrong")
 	errUnknownDataType         = errors.New("unknown shema DataType")
+	errContext                 = errors.New("context done or timeout")
 )
 
 type iterator = storage.Iterator
 
 type compactor interface {
+	add()
+	done()
```
Comment (congqixia) [inline internal/datanode/compactor.go:53]: logic ok, but `add` and `done` is not meaningful name of function. Maybe `start(runningState)`/ `complete(completedState)` is better?

File: internal/datanode/compaction_executor.go (hunk 2 above)
Comment (XuanYang-cn) [reply internal/datanode/compaction_executor.go:114]: Yes, this should stay LoadAndDelete

**Review by congqixia — APPROVED** [review]
/lgtm

**Pull request merged**